[HDR] Avoid colorSpace transformation and double caching when sharing an image with GPUProcess

shallawa opened the pull request:
#### 1cfbc06323c83c1436ab9208965a2665fc93ff73
<pre>
[HDR] Avoid colorSpace transformation and double caching when sharing an image with GPUProcess
<a href="https://bugs.webkit.org/show_bug.cgi?id=293450">https://bugs.webkit.org/show_bug.cgi?id=293450</a>
<a href="https://rdar.apple.com/151876825">rdar://151876825</a>

Reviewed by Simon Fraser.

Currently ShareableBitmap shares an image with GPUProcess by drawing it to a
SharedMemory only if its colorSpace is not sRGB. Otherwise it uses the data
provider to copy the pixels to a SharedMemory. In addition to the SharedMemory,
the image metadata which is stored in ShareableBitmapConfiguration, GPUProcess
can construct the original image.

Also NativeImage::draw() does the tone-mapping to SDR in WebProcess before
sharing the image with GPUProcess. It does this by drawing the image to an SDR
unaccelerated ImageBuffer. But this forces RemoteResourceCacheProxy to double
cache the image: one copy for the SDR and the other copy for the HDR.

These two steps can be avoided if
(1) the data provider approach is used for any colorSpace.
(2) the system tone-mapping APIs are used to tone-map the HDR image to SDR.
(3) the extended display ColorSpace is used for HDR layers.

* LayoutTests/fast/images/hdr-background-image.html:
* LayoutTests/fast/images/hdr-basic-image-dynamic-range-limit.html:
* LayoutTests/fast/images/hdr-basic-image.html:
* LayoutTests/fast/images/hdr-border-image.html:
* LayoutTests/fast/images/hdr-image-layer-sdr-dynamic-range-limit.html:
* LayoutTests/fast/images/hdr-pseudo-before-image.html:
* LayoutTests/fast/images/hdr-svg-inline-image.html:
The pixel tolerance in the ref test has to be increased due to using the
extended display ColorSpace for HDR layers. This should be revisited if
the system behavior changes.

* LayoutTests/platform/ios/compositing/hdr/hdr-basic-image-dynamic-range-limit-expected.txt:
* LayoutTests/platform/mac-wk2/compositing/hdr/hdr-basic-image-dynamic-range-limit-expected.txt:
HDR images with &quot;dynamic-range-limit: standard;&quot; have to be drawn to
HDR layers. Tone-mapping functions will be used to get the SDR images.

* Source/WebCore/platform/graphics/DestinationColorSpace.cpp:
(WebCore::DestinationColorSpace::asExtended const):
* Source/WebCore/platform/graphics/DestinationColorSpace.h:
* Source/WebCore/platform/graphics/ShareableBitmap.cpp:
(WebCore::ShareableBitmapConfiguration::ShareableBitmapConfiguration):
* Source/WebCore/platform/graphics/ShareableBitmap.h:
(WebCore::ShareableBitmapConfiguration::headroom const):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::setCGDynamicRangeLimitForImage):
(WebCore::GraphicsContextCG::drawNativeImageInternal):
* Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
(WebCore::NativeImage::draw):
* Source/WebCore/platform/graphics/cg/ShareableBitmapCG.mm:
(WebCore::ShareableBitmapConfiguration::ShareableBitmapConfiguration):
(WebCore::ShareableBitmap::createFromImagePixels):
(WebCore::ShareableBitmap::createCGImage const):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::recordNativeImageUse):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::displayColorSpace const):

Canonical link: <a href="https://commits.webkit.org/295899@main">https://commits.webkit.org/295899@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/85e9de1278385f5b117816987e1a66b9e2d6d5dd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/106473 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/26222 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16620 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111675 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/57069 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/108512 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/26889 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34725 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80864 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/109477 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/21322 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96082 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61194 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/20787 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/14183 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/56510 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90667 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/14218 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/114535 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33611 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24780 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89935 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33975 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/92313 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89642 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22877 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34535 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/12355 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/29220 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33536 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/38948 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/33282 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36635 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34880 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
```diff
@@ -281,6 +281,27 @@ void GraphicsContextCG::restore(GraphicsContextState::Purpose purpose)
+    RetainPtr toneMappingOptions = adoptCF(CFDictionaryCreate(kCFAllocatorDefault, toneMappingKeys, toneMappingValues, sizeof(toneMappingKeys) / sizeof(toneMappingKeys[0]), &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks));
+
+    CGContentToneMappingInfo toneMappingInfo = { kCGToneMappingReferenceWhiteBased, toneMappingOptions.get() };
+    CGContextSetContentToneMappingInfo(context, toneMappingInfo);
```
Is this something that is part of the CGContexts gstate and therefore part of the save/restore system?

shallawa replied:
Yes. And this is why at the end of this function we do:
```
    if (!stateSaver.didSave()) {
        ...    
        CGContextSetContentToneMappingInfo(context, oldToneMappingInfo);
        ...
    }
```

smfr replied:
Sam was maybe asking if we need to plumb `EDRTargetHeadroom` and `ContentToneMappingInfo` through our own GraphicsContextState, since context.save() and context.restore() won't work if we do. It's probably OK to not do that now since their usage is limited to just `GraphicsContextCG::drawNativeImageInternal()`, but if we need them for HDR colors in future, maybe we should.

I think `setCGDynamicRangeLimitForImage` could be a lambda in `GraphicsContextCG::drawNativeImageInternal()` to make it even more clear that the usage is limited.

shallawa replied:
Yes this is correct. If we add

```
GraphicsContext::setEDRTargetHeadroom()
GraphicsContext::setContentToneMappingInfo()
```

Then we will have to add the `EDRTargetHeadroom` and `ContentToneMappingInfo` through our own `GraphicsContextState`

smfr commented on Source/WebCore/platform/graphics/cg/ShareableBitmapCG.mm:
```diff
@@ -244,7 +249,13 @@
+#endif
+
     return adoptCF(CGImageCreate(size().width(), size().height(), bitsPerPixel / 4, bitsPerPixel, bytesPerRow, m_configuration.platformColorSpace(), m_configuration.bitmapInfo(), dataProvider, 0, shouldInterpolate == ShouldInterpolate::Yes, kCGRenderingIntentDefault));
+
```
Extra blank line here.

shallawa committed:
[HDR] Avoid colorSpace transformation and double caching when sharing an image with GPUProcess
https://bugs.webkit.org/show_bug.cgi?id=293450
rdar://151876825

Reviewed by Simon Fraser.

Currently ShareableBitmap shares an image with GPUProcess by drawing it to a
SharedMemory only if its colorSpace is not sRGB. Otherwise it uses the data
provider to copy the pixels to a SharedMemory. In addition to the SharedMemory,
the image metadata which is stored in ShareableBitmapConfiguration, GPUProcess
can construct the original image.

Also NativeImage::draw() does the tone-mapping to SDR in WebProcess before
sharing the image with GPUProcess. It does this by drawing the image to an SDR
unaccelerated ImageBuffer. But this forces RemoteResourceCacheProxy to double
cache the image: one copy for the SDR and the other copy for the HDR.

These two steps can be avoided if
(1) the data provider approach is used for any colorSpace.
(2) the system tone-mapping APIs are used to tone-map the HDR image to SDR.
(3) the extended display ColorSpace is used for HDR layers.

* LayoutTests/fast/images/hdr-background-image.html:
* LayoutTests/fast/images/hdr-basic-image-dynamic-range-limit.html:
* LayoutTests/fast/images/hdr-basic-image.html:
* LayoutTests/fast/images/hdr-border-image.html:
* LayoutTests/fast/images/hdr-image-layer-sdr-dynamic-range-limit.html:
* LayoutTests/fast/images/hdr-pseudo-before-image.html:
* LayoutTests/fast/images/hdr-svg-inline-image.html:
The pixel tolerance in the ref test has to be increased due to using the
extended display ColorSpace for HDR layers. This should be revisited if
the system behavior changes.

* LayoutTests/platform/ios/compositing/hdr/hdr-basic-image-dynamic-range-limit-expected.txt:
* LayoutTests/platform/mac-wk2/compositing/hdr/hdr-basic-image-dynamic-range-limit-expected.txt:
HDR images with "dynamic-range-limit: standard;" have to be drawn to
HDR layers. Tone-mapping functions will be used to get the SDR images.

* Source/WebCore/platform/graphics/DestinationColorSpace.cpp:
(WebCore::DestinationColorSpace::asExtended const):
* Source/WebCore/platform/graphics/DestinationColorSpace.h:
* Source/WebCore/platform/graphics/ShareableBitmap.cpp:
(WebCore::ShareableBitmapConfiguration::ShareableBitmapConfiguration):
* Source/WebCore/platform/graphics/ShareableBitmap.h:
(WebCore::ShareableBitmapConfiguration::headroom const):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::setCGDynamicRangeLimitForImage):
(WebCore::GraphicsContextCG::drawNativeImageInternal):
* Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
(WebCore::NativeImage::draw):
* Source/WebCore/platform/graphics/cg/ShareableBitmapCG.mm:
(WebCore::ShareableBitmapConfiguration::ShareableBitmapConfiguration):
(WebCore::ShareableBitmap::createFromImagePixels):
(WebCore::ShareableBitmap::createCGImage const):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::recordNativeImageUse):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::displayColorSpace const):

Canonical link: https://commits.webkit.org/295899@main
